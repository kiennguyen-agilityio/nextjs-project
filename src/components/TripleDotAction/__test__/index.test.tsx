import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import TripleDotActions from '..';

describe('TripleDotActions', () => {
  const mockToggleDropdown = jest.fn();
  const mockOnBlur = jest.fn();
  const mockOnDelete = jest.fn();
  const link = '/edit-link';

  let _handleClickOutside: (event: MouseEvent) => void;

  beforeEach(() => {
    jest.clearAllMocks();
    // Mock addEventListener and removeEventListener
    jest
      .spyOn(document, 'addEventListener')
      .mockImplementation((event, handler) => {
        if (event === 'mousedown') {
          _handleClickOutside = handler as (event: MouseEvent) => void;
        }
      });
    jest.spyOn(document, 'removeEventListener');
  });

  it('renders the TripleDot button', () => {
    render(
      <TripleDotActions
        isDropdownOpen={false}
        link={link}
        toggleDropdown={mockToggleDropdown}
        onBlur={mockOnBlur}
        onDelete={mockOnDelete}
      />,
    );

    expect(screen.getByTestId('triple-dot')).toBeInTheDocument();
  });

  it('opens the dropdown menu when the TripleDot button is clicked', () => {
    render(
      <TripleDotActions
        isDropdownOpen={false}
        link={link}
        toggleDropdown={mockToggleDropdown}
        onBlur={mockOnBlur}
        onDelete={mockOnDelete}
      />,
    );

    fireEvent.click(screen.getByTestId('triple-dot'));

    expect(mockToggleDropdown).toHaveBeenCalled();
    expect(document.addEventListener).toHaveBeenCalledWith(
      'mousedown',
      expect.any(Function),
    );
  });

  it('navigates to the edit link when the Edit button is clicked', () => {
    render(
      <TripleDotActions
        isDropdownOpen={true}
        link={link}
        toggleDropdown={mockToggleDropdown}
        onBlur={mockOnBlur}
        onDelete={mockOnDelete}
      />,
    );

    const editLink = screen.getByTestId('edit-link');
    expect(editLink).toHaveAttribute('href', link);

    fireEvent.click(editLink);

    expect(mockOnBlur).toHaveBeenCalled();
  });

  it('calls onDelete when the Delete button is clicked', () => {
    render(
      <TripleDotActions
        isDropdownOpen={true}
        link={link}
        toggleDropdown={mockToggleDropdown}
        onBlur={mockOnBlur}
        onDelete={mockOnDelete}
      />,
    );

    const deleteButton = screen.getByTestId('delete-button');
    fireEvent.click(deleteButton);

    expect(mockOnDelete).toHaveBeenCalled();
  });

  // Snapshot test
  it('matches the snapshot', () => {
    const { asFragment } = render(
      <TripleDotActions
        isDropdownOpen={false}
        link={link}
        toggleDropdown={mockToggleDropdown}
        onBlur={mockOnBlur}
        onDelete={mockOnDelete}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
