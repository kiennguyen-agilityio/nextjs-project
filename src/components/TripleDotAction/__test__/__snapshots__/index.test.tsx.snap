// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TripleDotActions matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <button
      aria-label="More options"
      class="text-gray-500 p-1 rounded-full hover:bg-gray-200"
      data-testid="triple-dot"
    >
      <svg
        class="w-6 h-6"
        fill="none"
        stroke="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h.01M12 12h.01M19 12h.01"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="4"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
